### ISSUE TITLE
Library fails to load on Windows x64 under SBCL 2.1.10 and CCL 1.12.1.

### ISSUE THREAD

**Issue by HanshenWang** [hanshuebner/bknr-datastore#21]
Thank you for taking the time to open source this library! I did first hear about it from 'Common Lisp Recipes (2016)'. 

It was smooth sailing until it came time to deploy. I develop with SBCL x64 on WSL Linux so it was my CD pipeline that caught this. The exact same error has occurred in the past as per this thread: https://old.reddit.com/r/Common_Lisp/comments/54kgtv/bknrdatastore_problems_in_windows_10/
```
;     Symbol "FSYNC" not found in the SB-POSIX package.
;   
;       Line: 501, Column: 17, File-Position: 20103
```
Just to summarize, /u/flaming_bird pointed out that SBCL defines FSYNC only under non-win32 systems (https://github.com/sbcl/sbcl/blob/a833d4df8373d42ea359ba0e667cb56ed66b55fc/contrib/sb-posix/interface.lisp#L179). While /u/knobo speculates using FlushFileBuffers instead of fsync is the right thing on windows.
```
#+win32
 (defun flush-file-buffers (fd)
   (sb-alien:alien-funcall
    (sb-alien:extern-alien "FlushFileBuffers"
                           (function (boolean) sb-alien:unsigned-long))
 fd))
```
I saw that the code in question used a conditional for SBCL here,
```
(defun fsync (stream)
  ;; FINISH-OUTPUT macht leider auch nichts anderes als FORCE-OUTPUT,
  ;; dabei waere sync()-Semantik zu erwarten.
  (finish-output stream)
  #+cmu
  (unix:unix-fsync (kernel::fd-stream-fd stream))
  #+sbcl
  (sb-posix:fsync (sb-kernel::fd-stream-fd stream)))
```
and I have to confess here I'm a non-compsci major dabbling around in Lisp, so I'm way outta depth, but am glad to offer any help testing changes. I did however at this point try CCL to see if that implementation has better support for windows. The windows runner on Github Actions threw this back at me when loading the library:

```
Read error between positions 28843 and 29607 in D:/a/project-isidore/project-isidore/buildcache/quicklisp/dists/quicklisp/software/bknr-datastore-20191227-git/src/data/txn.lisp.
[package bknr.datastore.tests]................
> Error: Foreign function not found: WIN64::|truncate|
> Type :GO to continue, :POP to abort, :R for a list of available restarts.
> While executing: CCL::LOAD-EXTERNAL-FUNCTION, in process listener(1).
> If continued: Skip loading "D:\\a\\project-isidore\\project-isidore\\bin\\make.lisp"
> Type :? for other options.
```

Now it's been a real pleasure using this library on my development platform, but I'm kind of at a loss as to what to do for windows support. I understand lispers consider windows to be the red headed stepchild of OS's out there, and this is just a hobby project, so I understand it may be some time before a fix. Any help is appreciated!

Cheers,

**Comment by hanshuebner** [comment]
The call to `fsync` is optional in the sense that it won't make any functional difference if you don't call it.  The idea behind it is that after `fsync` has been invoked, all file updates have been written to stable storage, which is not quite true even on Linux anyway.

The call to `truncate` is required when restoring from the log to a certain point in time or when there is an error during restore.  It is not needed during normal operation.

Thus, you can probably solve both problems by commenting out the code in question.  You'll lose the ability to do point-in-time restores and automatic restores will become less robust.  I understand that these might not be concerns for you.

**Comment by HanshenWang** [comment]
Ah very clear explanation thank you. I got things working for SBCL with  `#+(and sbcl (not win32))` in the locations mentioned. Shall I put together a tiny pull request? Or is it a more sensible default to leave the options as is?

Thanks again for the library, particularly the datastore-manual.pdf docs.

Best,

**Comment by Jaso-N7** [comment]
Hi @hanshuebner , I had this exact issue on Windows 10. While I am currently stuck with Windows for development, I am creating a solution for production that will need to be portable, and will need the features of restoring from the log and would also need to test this functionality during development.

Taking your suggestion into account, I was able to modify the code and got it loading via quicklisp without any issues. Here's the modification:

txn.lisp L728
`  #+(and :windows :ccl)
  (ccl:with-cstrs ((filename (namestring pathname)))
    (#_fopen filename (ccl::make-cstring "w")))
  #+(and :linux openmcl)
  (ccl:with-cstrs ((filename (namestring pathname)))
    (#_truncate filename position))`

Of course my macros would need some tidying up, however using FOPEN with the "W" mode accomplishes file truncation. 

Should I create a pull request for this?

**Comment by hanshuebner** [comment]
> Should I create a pull request for this?
 
I don't think I want to see this silently commented out.  Rather, the point-in-time-restore feature should be disabled on Windows and the documentation should clearly say so.

**Comment by Jaso-N7** [comment]
> I don't think I want to see this silently commented out. Rather, the point-in-time-restore feature should be disabled on Windows and the documentation should clearly say so.

Hi Hans, thanks for responding. Not sure what you meant by this. I thought that using `fopen` was a good alternative to `truncate` on Windows systems. The macros just clearly showed a specific case for that OS.

I went through the tutorial and was able to do a point-in-time restore on Windows, at least that what I thought it was, where I chose a timestamp and restored up to that point. Are you saying that this should not be the case? Unless I am confusing the terminology.

**Comment by hanshuebner** [comment]
Does `fopen` truncate the file?

**Comment by Jaso-N7** [comment]
> Does `fopen` truncate the file?

Yes, I tested it on a simple text file before applying it in the code. You can try it with this CCL FFI:

```
(ccl:with-cstrs ((filename (namestring #p"c:/tmp/truncated.txt")))
  (#_fopen filename (ccl::make-cstring "w")))
```

Just enter any text you want in `truncated.txt`, run that code from the REPL and opening the file again will show that it's blank.

`(lisp-implementation-version)` => "Version 1.12 (v1.12) WindowsX8664"

**Comment by hanshuebner** [comment]
OK, so it truncates the file to be empty, whereas `truncate` truncates the file to the provided `position`.  This is quite the difference.

**Comment by Jaso-N7** [comment]
Ahh, that makes a world of difference! I was able to find this C function [_chsize](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/chsize?view=msvc-170).

Excerpt from the link: 

> The _chsize function extends or truncates the file associated with fd to the length specified by size. 

It seems to behave the same way as `truncate`.  I replaced the `fopen` snippet with `_chsize`, from my previous example:

```
#+(and :windows :ccl)
(ccl:with-cstrs ((filename (namestring pathname)))
    (#__chsize filename position))
```
Would this suffice? Unfortunately, my FFI skills are truly subpar and I am unable to produce a working standalone example, as it keeps complaining about the filename not being of type `(signed-byte 32)`

**Comment by hanshuebner** [comment]
`_chsize` (or `_chsize_s`) looks good, but I'm unable to help with getting it to run as I don't have a Windows development environment ready to go.

**Comment by Jaso-N7** [comment]
Thanks, I understand and appreciate the responses / assistance. I'll keep `_chsize` in my quickload(ed) version and let you know if there are any breaking changes or not.

The reason I keep trying with Windows is more for deploying to clients production environments where they refuse to entertain the idea that Unix exist.

**Comment by Jaso-N7** [comment]
Just an update since using `_chsize`. When the log is being truncated, this is the error that is being reported:

`The value #<A Foreign Pointer [stack-allocated] #x258AF180> is not of the expected type (SIGNED-BYTE 32)` 

This happens during the restore. The error could be referencing either file name (file descriptor) or the size (or both) parameters being passed to `_chsize`

**Comment by hanshuebner** [comment]
The `_chsize` function accepts a file descriptor, not a file name, as the
`fd` argument.  You will have to open the log file before you can truncate
it.

Am Do., 30. Juni 2022 um 02:02 Uhr schrieb Jason Robinson <
***@***.***>:

**Comment by Jaso-N7** [comment]
Thanks for pointing that out! My C has completed oxidized :)
I will try it with an open file.

**Comment by Jaso-N7** [comment]
Hi Hans, thanks for all the assistance thus far and for work done on this amazing system. Currently, I will have to level up my skillset to pull it off for Windows. It is as you previously stated, I will have to workaround/disable the point-in-time restore until then.
